source/extensions/filters/network/thrift_proxy/conn_manager.cc

```diff
@@ -464,22 +631,39 @@ void ConnectionManager::ActiveRpc::finalizeRequest() {
 
 bool ConnectionManager::ActiveRpc::passthroughSupported() const {
   for (auto& entry : decoder_filters_) {
-    if (!entry->handle_->passthroughSupported()) {
+    if (!entry->decoder_handle_->passthroughSupported()) {
+      return false;
+    }
+  }
+  for (auto& entry : encoder_filters_) {
+    if (!entry->encoder_handle_->passthroughSupported()) {
       return false;
     }
   }
   return true;
 }
 
-FilterStatus ConnectionManager::ActiveRpc::passthroughData(Buffer::Instance& data) {
-  passthrough_ = true;
+void ConnectionManager::ActiveRpc::passthroughDataSetup(Buffer::Instance& data) {
   filter_context_ = &data;
   filter_action_ = [this](DecoderEventHandler* filter) -> FilterStatus {
     Buffer::Instance* data = absl::any_cast<Buffer::Instance*>(filter_context_);
     return filter->passthroughData(*data);
   };
+}
+
+FilterStatus ConnectionManager::ActiveRpc::passthroughData(Buffer::Instance& data) {
+  passthrough_ = true;
+  passthroughDataSetup(data);
+
+  return applyDecoderFilters();
+}
 
-  return applyDecoderFilters(nullptr);
+void ConnectionManager::ActiveRpc::messageBeginSetup(MessageMetadataSharedPtr metadata) {
+  filter_context_ = metadata;
+  filter_action_ = [this](DecoderEventHandler* filter) -> FilterStatus {
+    MessageMetadataSharedPtr metadata = absl::any_cast<MessageMetadataSharedPtr>(filter_context_);
+    return filter->messageBegin(metadata);
```

Comment:
What do you think of replacing all of these `*Setup()` method with a generic one? E.g.:

```
void prepareFilterAction(absl::any data, DecodeEventHandlerState state) {
   filter_context_ = std::move(data);
   switch (state) {
      case DecodeEventHandlerState::MessageBegin:
         filter_action_ = [this](DecoderEventHandler* filter) -> FilterStatus {
           MessageMetadataSharedPtr metadata = absl::any_cast<MessageMetadataSharedPtr>(filter_context_);
           return filter->messageBegin(metadata);            
          };
          break;
       ....
   }
}

```

For which we'd add an enum class next to https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/network/thrift_proxy/decoder_events.h#L20, e.g.:

```
enum class DecoderEventHandler {
  MessageBegin,
  ....
};
```

It would be a long switch statement, but it'll make the change less noisy (I think) and we'd have everything in one place. 

We can even take it one step further and have another helper with a 3rd parameter for actually running the filters:

```
FilterStatus runFilters(absl::any data, DecodeEventHandlerState state, std::function< FilterStatus()> run_filters_fn) {
   prepareFilterAction(data, state);
   return run_filters_fn();
}
```

which you can conveniently call like this:

```
FilterStatus ConnectionManager::ActiveRpc::passthroughData(Buffer::Instance& data) {
  passthrough_ = true;
  return runFilters(data, DecoderEventHandlerState::PassthroughData, [this]() -> FilterStatus {  return applyDecoderFilters(); });
}
```

What do you think? This should DRY things up quite a bit. #17930 might serve as inspiration.

Reply:
I'll vote for DRY so yes.